src/dataframe/dataframe.jl

```diff
@@ -1176,30 +1225,79 @@ Base.convert(::Type{DataFrame}, A::AbstractMatrix) = DataFrame(A)
 Base.convert(::Type{DataFrame}, d::AbstractDict) = DataFrame(d, copycols=false)
 
 function Base.push!(df::DataFrame, row::Union{AbstractDict, NamedTuple}; cols::Symbol=:setequal,
-                    columns::Union{Nothing,Symbol}=nothing)
+                    columns::Union{Nothing,Symbol}=nothing, promote::Bool=(cols in [:union, :subset]))
     if columns !== nothing
         cols = columns
         Base.depwarn("`columns` keyword argument is deprecated. Use `cols` instead. ", :push!)
     end
-    possible_cols = (:orderequal, :setequal, :intersect, :subset)
+    possible_cols = (:orderequal, :setequal, :intersect, :subset, :union)
     if !(cols in possible_cols)
         throw(ArgumentError("`cols` keyword argument must be any of :" * join(possible_cols, ", :")))
     end
+
     nrows, ncols = size(df)
     targetrows = nrows + 1
+
     if ncols == 0 && row isa NamedTuple
         for (n, v) in pairs(row)
             setproperty!(df, n, fill!(Tables.allocatecolumn(typeof(v), 1), v))
         end
         return df
     end
+
+    if cols == :union
+        if !all(x -> x isa Symbol, keys(row))
+            throw(ArgumentError("when `cols=:union all keys of row must be Symbol"))
+        end
+        for (i, colname) in enumerate(_names(df))
+            col = _columns(df)[i]
+            if haskey(row, colname)
+                val = row[colname]
+            else
+                val = missing
+            end
+            S = typeof(val)
+            T = eltype(col)
+            V = promote_type(S, T)
+            if S <: T || V <: T || !promote
```

Comment:
I wonder whether it wouldn't be better for performance to avoid calling `promote_type` if `S <: T || !promote` (for type-unstable cases). Same for `append!`, though it should be less of a concern there since you add multiple rows.

Also in some cases maybe having something like `push_helper` could make sense?

Reply:
OK - fixed this by reordering conditions.

I do not see where `push_helper` could help as we are processing each column exactly once.

Reply:
We have to call `promote_type` twice now but I think it is OK, as we call it for the second time only if we allocate a column. Also maybe compiler will be smart enough to optimize this out.